compiler/optimizer/parallelize.go

```diff
@@ -2,189 +2,211 @@ package optimizer
 	}
-	egressSortKey, err := o.sortKeyOfTrunk(trunk)
+	srcSortKey, err := o.sortKeyOfSource(scan)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	if len(egressSortKey.Keys) > 1 {
-		// XXX don't yet support multi-key ordering
-		return nil
+	if len(srcSortKey.Keys) > 1 {
+		// XXX Don't yet support multi-key ordering.  See Issue #2657.
+		return nil, nil
 	}
-	// This logic requires that there is only one trunk in the From,
-	// as checked above.
-	sortKey, err := o.sortKeyOfSource(trunk.Source, order.Nil)
+	// concurrentPath will check that the path consisting of the original source
+	// sequence and any lifted sequence is still parallelizable.
+	n, outputKey, _, needMerge, err := o.concurrentPath(ops[1:], srcSortKey)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	inlineSequentials(seq)
-	// Check that the path consisting of the original from
-	// sequence and any lifted sequence is still parallelizable.
-	if trunk.Seq != nil && len(trunk.Seq.Ops) > 0 {
-		n, newSortKey, err := o.splittablePath(trunk.Seq.Ops, sortKey)
-		if err != nil {
-			return err
-		}
-		if n != len(trunk.Seq.Ops) {
-			return nil
+	// XXX Fix this to handle multi-key merge. See Issue #2657.
+	if len(outputKey.Keys) > 1 {
+		return nil, nil
+	}
+	head := ops[:n+1]
+	tail := ops[n+1:]
+	par := &dag.Parallel{Kind: "Parallel", Any: true}
+	for k := 0; k < replicas; k++ {
+		par.Ops = append(par.Ops, &dag.Sequential{Kind: "Sequential", Ops: copyOps(head)})
+	}
+	var merge dag.Op
+	if needMerge {
+		// At this point, we always insert a merge as we don't know if the
+		// downstream DAG requires the sort order.  A later step will look at
+		// the fanin from this parallel structure and see if the merge can be
+		// removed while also pushing additional ops from the output segment up into
+		// the parallel branches to enhance concurrency.
+		merge = &dag.Merge{
+			Kind:  "Merge",
+			Expr:  &dag.This{Kind: "This", Path: outputKey.Primary()},
+			Order: outputKey.Order,
 		}
-		// If the trunk operators affect the scan sort key, then update
-		// it here so the merge will properly happen below...
-		sortKey = newSortKey
+	} else {
+		merge = &dag.Combine{Kind: "Combine"}
 	}
-	if len(seq.Ops) < 2 {
-		// There are no operators past the trunk.  Just parallelize
-		// and merge the trunk here.
-		if err := insertMerge(seq, sortKey); err != nil {
-			return err
+	return append([]dag.Op{par, merge}, tail...), nil
+}
+
+func (o *Optimizer) findParPullups(op dag.Op) {
+	walkOp(op, false, func(op dag.Op) {
+		if seq, ok := op.(*dag.Sequential); ok {
+			for ops := seq.Ops; len(ops) >= 3; ops = ops[1:] {
+				o.parPullUp(ops)
+			}
 		}
-		replicateTrunk(from, trunk, replicas)
-		return nil
+	})
+}
+
+// parPullUp examines a sequence of Ops to see if there's an opportunity to
+// lift operations downstream from a parallel Op into the parallel paths to
+// enhance concurrency.  If so, we modify the sequential ops in place.
+func (o *Optimizer) parPullUp(ops []dag.Op) {
```

Comment:
Maybe call this liftParPullups.
```suggestion
// liftParPullups examines a sequence of Ops to see if there's an opportunity to
// lift operations downstream from a parallel Op into the parallel paths to
// enhance concurrency.  If so, we modify the sequential ops in place.
func (o *Optimizer) liftParPullups(ops []dag.Op) {
```

Reply:
I landed on `optimizeParallels` and `liftIntoParPaths`